pkg/controller/support/credentials/cloudauth/gcr/gcr_test.go

```diff
@@ -1,102 +1,321 @@
-	os.Unsetenv(envCredentials)
-	t.Cleanup(func() {
-		os.Setenv(envCredentials, secret)
-	})
-
-	registry := &cloudauth.Registry{}
-	err := Register(context.TODO(), logr.Discard(), registry)
-	if err != nil {
-		t.Error(err)
-	}
+type roundTripFunc func(r *http.Request) (*http.Response, error)
+
+func (s roundTripFunc) RoundTrip(r *http.Request) (*http.Response, error) {
+	return s(r)
 }
 
 func TestAuthenticate(t *testing.T) {
-	if os.Getenv(envCredentials) == "" {
-		t.Skip("Skipping, gcp not setup")
-	}
+	defaultCtx := context.Background()
 
-	p, err := newProvider(context.TODO(), logr.Discard())
-	if err != nil {
-		t.Fatal(err)
-	}
+	// Canceled context for testing failed do request
+	canceledCtx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	// logger for comparing expected vs actual logging.
+	observerCore, observedLogs := observer.New(zap.DebugLevel)
+	log := zapr.NewLogger(zap.New(observerCore))
 
-	ctx := context.Background()
-	t.Run("invalid url", func(t *testing.T) {
-		auth, err := p.authenticate(ctx, "bogus.g.io")
-		if err == nil {
-			t.Error("unexpected error")
-		}
-		if auth != nil {
-			t.Errorf("auth not nil")
-		}
-	})
-
-	t.Run("valid url", func(t *testing.T) {
-		for _, tt := range []struct {
-			name string
-		}{
-			{"gcr.io"},
-			{"us-west1-docker.pkg.dev"},
-		} {
-			t.Run(tt.name, func(t *testing.T) {
-				auth, err := p.authenticate(ctx, tt.name)
+	// test server
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "New Server, %s", r.Proto)
+	}))
+	ts.Close()
+
+	// test server IP and port information for error logging
+	tsAddress := ts.Listener.Addr()
+
+	// expected errors
+	invalidUrlErr := errors.New(fmt.Sprintf("invalid gcr url test-ts should match %s", gcrRegex))
+	gcrTokenAccessErr := errors.New("unable to access gcr token from oauth: default error message")
+	gcrRegistryErr := errors.New("GCR registry \"https://hi-docker.pkg.dev\" is unusable: default error message")
+	ctxCanceledErr := errors.New(fmt.Sprintf("Get \"http://oauth2accesstoken:***@%s?client_id=forge&service=serviceName\": context canceled", tsAddress.String()))
+	non200StatusErr := errors.New("failed to obtain token, received unexpected response code: 400\nresponse: \"nope\"")
+	noResTokenErr := errors.New("failed, no gcr token in bearer response:\n{\"token\":\"\",\"access_token\":\"\",\"refresh_token\":\"\"}")
+
+	for _, tt := range []struct {
+		name                        string
+		serverName                  string
+		ctx                         context.Context
+		roundTripper                roundTripFunc
+		provider                    *gcrProvider
+		authConfig                  *types.AuthConfig
+		defaultChallengeLoginServer cloudauth.LoginChallenger
+		expectedLogMessage          string
+		expectedError               error
+	}{
+		{
+			"invalid-url",
+			"test-ts",
+			defaultCtx,
+			nil,
+			&gcrProvider{
+				logger:      log,
+				tokenSource: nil,
+			},
```

Comment:
This `*gcrProvider` is built for each test such that I think a `fakeGCRProvider(...)` func would be useful.

Reply:
I was using the same logger and only really changing the value of the token source erring in the provider, so I updated it to this. LMK what you think. 

https://github.com/dominodatalab/hephaestus/pull/67/files#diff-5a64af87db87fed02f741d838e0256f07ac212ebbc0eea7269d969b6046cdb3cR62